execute/report/report.go

```diff
@@ -191,19 +200,68 @@ func (b *execReportBuilder) checkMessage(
 		return execReport, TokenDataFetchError, nil
 	}
 
-	execReport.TokenData = padSlice(execReport.TokenData, idx+1, nil)
-	execReport.TokenData[idx] = tokenData
+	result.TokenData = padSlice(execReport.TokenData, idx+1, nil)
+	result.TokenData[idx] = tokenData
 	b.lggr.Infow(
 		"read token data",
 		"messageID", msg.Header.MessageID,
 		"sourceChain", execReport.SourceChain,
 		"seqNum", msg.Header.SequenceNumber,
 		"data", tokenData)
 
-	// TODO: Check for valid nonce
+	// 3. Check if the message has a valid nonce.
+	if msg.Header.Nonce != 0 && b.nonceCheckingEnabled {
+		// Sequenced messages have non-zero nonces.
+
+		if _, ok := b.sendersNonce[execReport.SourceChain]; !ok {
+			b.lggr.Errorw("Skipping message - nonces not available for chain",
+				"messageID", msg.Header.MessageID,
+				"sourceChain", execReport.SourceChain,
+				"seqNum", msg.Header.SequenceNumber,
+			)
+			return execReport, MissingNoncesForChain, nil
```

Comment:
Would it be useful to `fmt.Errorf` the `SourceChain` here for the `MissingNoncesForChain` error? Where would this error should up?

Reply:
Just the logs. The only message state currently used used is to check if the message is ready.

Reply:
I'll add the message state to the structured log, that should be easier to key off of than the message text.